server/src/main/java/org/elasticsearch/common/logging/ESLogMessage.java

```diff
@@ -36,10 +36,13 @@
  */
 public class ESLogMessage extends MapMessage<ESLogMessage, Object> {
     private final List<Object> arguments = new ArrayList<>();
+    private String messagePattern;
 
     public ESLogMessage(String messagePattern, Object... args) {
```

Comment:
I think we don't wont this. In log4j you no longer have to use if(isDebug) guard style. THat means when someone calls 
```
logger.debug("some message {}", objectWithExpensiveToString)
```
it does not require to execute the toString and create an expensive string unless the debug level (or any other) is actually set.
The messagePattern and arguments getters are needed because we use the same object to format message to a header (different formatting then in logs)

For json logging I do similar think. I make sure that I store an object that lazily evaluates to "message" json field. https://github.com/elastic/elasticsearch/blob/master/server/src/main/java/org/elasticsearch/common/logging/ESLogMessage.java#L46

Reply:
> The messagePattern and arguments getters are needed because we use the same object to format message to a header (different formatting then in logs)

it is actually something that should be cleaned up

Reply:
I forgot about he lazy evaluation, thanks for the reminder.  You are right we probably don't want to eagerly resolve. We would have to if we wanted to use the message for key, but having an explicit key is fine for now (that has it has always worked) and in the future we may explore using he call site.